source/http_client_connection.c

```diff
@@ -0,0 +1,551 @@
+            aws_http_connection_close(http_connection->connection);
+        }
+
+        aws_http_connection_release(http_connection->connection);
+        http_connection->connection = NULL;
+    }
+
+    if (http_connection->shutdown_called) {
+        aws_mem_release(http_connection->allocator, http_connection);
+    }
+}
+
+static void s_on_client_connection_setup(struct aws_http_connection *connection, int error_code, void *user_data) {
+
+    struct py_http_connection *py_connection = user_data;
+    PyGILState_STATE state = PyGILState_Ensure();
+    PyObject *result = NULL;
+    PyObject *capsule = NULL;
+
+    if (!error_code) {
+        py_connection->connection = connection;
+        capsule =
+            PyCapsule_New(py_connection, s_capsule_name_http_client_connection, s_http_client_connection_destructor);
+        py_connection->capsule = capsule;
+    } else {
+        aws_mem_release(py_connection->allocator, py_connection);
+    }
+
+    result = PyObject_CallFunction(py_connection->on_connection_setup, "(Ni)", capsule, error_code);
+
+    Py_DECREF(py_connection->on_connection_setup);
+    Py_XDECREF(result);
+
+    PyGILState_Release(state);
+}
+
+static void s_on_client_connection_shutdown(struct aws_http_connection *connection, int error_code, void *user_data) {
+    (void)connection;
+    struct py_http_connection *py_connection = user_data;
+    py_connection->shutdown_called = true;
+
+    if (!py_connection->destructor_called) {
+        PyGILState_STATE state = PyGILState_Ensure();
+        PyObject *result = PyObject_CallFunction(py_connection->on_connection_shutdown, "(i)", error_code);
+        Py_XDECREF(result);
+        PyGILState_Release(state);
+    } else {
+        aws_mem_release(py_connection->allocator, py_connection);
+    }
+
+    Py_DECREF(py_connection->on_connection_shutdown);
+}
+
+PyObject *aws_py_http_client_connection_create(PyObject *self, PyObject *args) {
+    (void)self;
+
+    struct py_http_connection *py_connection = NULL;
+    struct aws_allocator *allocator = aws_crt_python_get_allocator();
+
+    PyObject *bootstrap_capsule = NULL;
+    PyObject *on_connection_shutdown = NULL;
+    PyObject *on_connection_setup = NULL;
+    const char *host_name = NULL;
+    Py_ssize_t host_name_len = 0;
+    uint16_t port_number = 0;
+    Py_ssize_t initial_window_size = PY_SSIZE_T_MAX;
+    PyObject *py_socket_options = NULL;
+    PyObject *tls_conn_options_capsule = NULL;
+
+    if (!PyArg_ParseTuple(
+            args,
+            "OOOs#HOO",
+            &bootstrap_capsule,
+            &on_connection_setup,
+            &on_connection_shutdown,
+            &host_name,
+            &host_name_len,
+            &port_number,
+            &py_socket_options,
+            &tls_conn_options_capsule)) {
+        PyErr_SetNone(PyExc_ValueError);
+        goto error;
+    }
+
+    if (!bootstrap_capsule || !PyCapsule_CheckExact(bootstrap_capsule)) {
+        PyErr_SetNone(PyExc_ValueError);
```

Comment:
I think you want ```PyErr_SetString(PyExc_ValueError, "bootstrap is invalid");``` for each of these arguments, otherwise this is very hard to diagnose for the user.